### DISCUSSION TITLE
Convert temperature difference into valve position

### DISCUSSION THREAD

**Discussion by nliaudat** [deklaus/OpenValveControl#4] [General]
Hi, 

I am very interested in your project as I have [mine](https://github.com/nliaudat/esp32_8ch_motor_shield) doing the same for 8 channels. 

I recommend you to change your temperature to valve position formula. 

I have experimented a lot and finally a [sigmoid like formula](https://github.com/nliaudat/floor-heating-controller/wiki/heating_channel_tanh) is the best I found.

I have better results with the sigmoid formula as it lower the time to cool the rooms. (I have a heat only underfloor heating with very good insulation a the building). 

For heat/cool underfloor heating or bad insulation, the [linear formula](https://github.com/nliaudat/floor-heating-controller/wiki/heating_channel_linear) makes the job

Adding a "min movement change"  valve position (5% is a good value) help to filter oscillations.

To conduct weather forecast, the simple is the best and you can introduce a "boost heating" factor which can be controlled by your rednode instance. 

Regards

PS. Your researches on BEMF in your user manual are great !!

**Comment by deklaus** [comment]
Thank you for this interesting advice on calculating the valve setpoint. I really appreciate your contribution. Your floor heating control project is great!

Indeed, I haven't paid much attention to temperature control, as my heating system already sets a weather-compensated flow temperature as described. That's why I currently use my controller more like a hydraulic balancing system and to switch individual valve zones on or off via the web interface, which is why the hardware only supports 4 valves.

My formula in the control via Node-RED is a deliberately simple example to demonstrate two-point control with hysteresis. There are certainly better algorithms for temperature control. Your formula can of course be integrated via Node-RED in a similar way, of course. 

However, I currently see the temperature control and also service functions such as a weekly reference run or descaling run at the higher control level, e.g. via smart home software or via Node-RED, and not in the “OVC” control unit.  But you can also see it differently and discuss it with each other. That's what github is for.

The control unit has been working very reliably and trouble-free for months now. However, there are still some important things to implement in the software (position estimation, error messages and data protection/security of WLAN access, etc.), which I am prioritizing. I would like to have these things done by the fall.

I look forward to any further suggestions and comments.
Regards

**Comment by nliaudat** [comment]
> However, I currently see the temperature control and also service functions such as a weekly reference run or descaling run at the higher control level, e.g. via smart home software or via Node-RED, and not in the “OVC” control unit

Me too. That's why I have a full remote controlled version per rest api or hassio.

I'll have a look at your progress on position estimation. My implementation of BEMF is bad cause with esphome firmware, I cannot have access easyly to low level ADC or PWM. 

If you want to have hardware low level integrated motor control, you can have a look at [esp-idf MCPWM](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/mcpwm.html)

For position control, I have not find a good way to do it cause pulse counting, hall sensors, etc loose position in case of restart. I prefer [IR opto solution](https://github.com/nliaudat/floor-heating-proportional-valve). 

Regards

**Reply by deklaus** [reply]
Of course, precise position evaluation is desirable. However, the inexpensive thermoelectric actuators even manage with a simple PWM control.
In order to keep the effort as low as possible, I currently only see end position or blocking detection via motor current and back EMF, even if this requires a regular reference run. In between, the drive moves in relative steps (closing when too warm and opening when too cold).
An incremental measuring system also requires occasional reference runs to correct any errors.
